great_expectations/core/config_provider.py

```diff
@@ -4,26 +4,45 @@
 import os
 from abc import ABC, abstractmethod
 from collections import OrderedDict
-from typing import Dict, Optional, Type, cast
+from typing import Any, Dict, Optional, Type, cast
 
+from great_expectations.core.config_substitutor import ConfigurationSubstitutor
 from great_expectations.core.yaml_handler import YAMLHandler
 from great_expectations.data_context.types.base import GeCloudConfig
-from great_expectations.data_context.util import (
-    substitute_all_config_variables,
-    substitute_config_variable,
-)
 
 yaml = YAMLHandler()
 
 
 class AbstractConfigurationProvider(ABC):
+    def __init__(self) -> None:
+        self._substitutor = ConfigurationSubstitutor()
+
     @abstractmethod
     def get_values(self) -> Dict[str, str]:
         """
         Retrieve any configuration variables relevant to the provider's environment.
         """
         pass
 
+    def substitute_config(
+        self, config: Any, config_values: Optional[Dict[str, str]] = None
+    ) -> Any:
+        """
+        Utilizes the underlying ConfigurationSubstitutor instance to substitute any
+        $VARIABLES with their corresponding config variable value.
+
+        Args:
+            config: The config object to update.
+            config_values: The dictionary of values to use during the substitution process.
+                           If omitted, any values derived from registered providers will be used.
+
+        Returns:
+            The input config object with any $VARIABLES replaced with their corresponding config values.
+        """
+        if config_values is None:
```

Comment:
I know we do this a lot throughout the codebase but I find the pattern of allowing a passed in value or acting on internal state confusing. I guess in the other examples internal state is mutated but that doesn't seem to be the case here. Is this to allow overriding of default values?

Reply:
We set this to `None` as we're not able to use default dictionaries (due to shared state with mutable defaults in Python).

In essence, we generally use the values derived from `self.get_values()` but also want the ability to take in runtime values - an example of this can be found in `yaml_config_validator.py` (where you've commented on the `update()` call).

Note that the default call to `self.get_values()` generates values JIT and therefore we're never working with internal state. Lmk if that makes sense.